File: VB/CustomItemExtension/CustomItems/SunburstChart/SunburstItemControlProvider.vb
Imports System
Imports System.Collections.Generic
Imports System.Drawing
Imports System.Linq
Imports System.Windows.Forms
Imports DevExpress.DashboardCommon
Imports DevExpress.DashboardCommon.ViewerData
Imports DevExpress.DashboardWin
Imports DevExpress.Skins
Imports DevExpress.TreeMap
Imports DevExpress.Utils
Imports DevExpress.Utils.Extensions
Imports DevExpress.XtraReports.UI
Imports DevExpress.XtraTreeMap

Namespace DevExpresss.DashboardWin.CustomItemExtension
	Public Class SunburstItemControlProvider
		Inherits CustomControlProviderBase

		Private skipSelectionEvent As Boolean = False
		Private sunburst As SunburstControl
		Private dataAdapter As SunburstFlatDataAdapter
		Private dashboardItem As CustomDashboardItem(Of SunburstItemMetadata)
		Private toolTipController As ToolTipController
		Private multiDimensionalData As MultiDimensionalData
		Private flatData As DashboardFlatDataSource
		Private emptyTitle As Title

		Protected Overrides ReadOnly Property Control() As Control
			Get
				Return sunburst
			End Get
		End Property
		Public Sub New(ByVal dashboardItem As CustomDashboardItem(Of SunburstItemMetadata))
			Me.dashboardItem = dashboardItem
			sunburst = New DashboardSunburstControl()
			sunburst.BorderOptions.Visible = False
			Me.dataAdapter = New SunburstFlatDataAdapter()
			sunburst.DataAdapter = dataAdapter
			toolTipController = New ToolTipController()
			AddHandler toolTipController.BeforeShow, AddressOf ToolTipController_BeforeShow
			sunburst.ToolTipController = toolTipController
			AddHandler sunburst.SelectionChanged, AddressOf Sunburst_SelectionChanged
			emptyTitle = New Title() With {.Visible = False}
			sunburst.Titles.Add(emptyTitle)
			AddHandler sunburst.MouseClick, AddressOf Sunburst_MouseClick
			AddHandler sunburst.LookAndFeel.StyleChanged, AddressOf SunburstLookAndFeelStyleChanged
			UpdateSunburstItemStyle()
		End Sub
		Protected Overrides Sub UpdateControl(ByVal customItemData As CustomItemData)
			ClearDataBindings()
			If ValidateBindings() Then
				flatData = customItemData.GetFlatData(New DashboardFlatDataSourceOptions() With {
					.AddColoringColumns = True,
					.AddDisplayTextColumns = True
				})
				multiDimensionalData = customItemData.GetMultiDimensionalData()
				SetDataBindings(flatData)
				SetColorizer(flatData)
				SetSelectionMode()
			End If
		End Sub
		Protected Overrides Sub SetSelection(ByVal selection As CustomItemSelection)
			skipSelectionEvent = True
			Dim selectedRows As IList(Of DashboardFlatDataSourceRow) = selection.GetDashboardFlatDataSourceRows(flatData)
			sunburst.SelectedItems.Clear()
			selectedRows.ForEach(Sub(r) sunburst.SelectedItems.Add(r))
			skipSelectionEvent = False
		End Sub
		Protected Overrides Function GetPrintableControl(ByVal customItemData As CustomItemData, ByVal exportInfo As CustomItemExportInfo) As XRControl
			Dim container As New PrintableComponentContainer()
			container.PrintableComponent = sunburst
			Return container
		End Function
		Private Sub ClearDataBindings()
			sunburst.Colorizer = Nothing
			dataAdapter.LabelDataMember = Nothing
			dataAdapter.ValueDataMember = dataAdapter.LabelDataMember
			dataAdapter.DataSource = dataAdapter.ValueDataMember
			dataAdapter.GroupDataMembers.Clear()
		End Sub
		Private Function ValidateBindings() As Boolean
			If Interactivity.IsDrillDownEnabled Then
				emptyTitle.Text = "The Sunburst Item does not support Drill-Down"
				emptyTitle.Visible = True
				Return False
			Else
				emptyTitle.Visible = False
			End If
			Return dashboardItem.Metadata.Value IsNot Nothing AndAlso dashboardItem.Metadata.Arguments.Count() > 0
		End Function
		Private Sub SetDataBindings(ByVal flatDataSource As DashboardFlatDataSource)
			dataAdapter.ValueDataMember = dashboardItem.Metadata.Value.UniqueId
			dataAdapter.LabelDataMember = flatDataSource.GetDisplayTextColumn(dashboardItem.Metadata.Arguments.Last().UniqueId).Name
			dataAdapter.GroupDataMembers.AddRange(dashboardItem.Metadata.Arguments.Where(Function(d) d IsNot dashboardItem.Metadata.Arguments.Last()).Select(Function(d) flatDataSource.GetDisplayTextColumn(d.UniqueId).Name).ToList())
			Try
				dataAdapter.DataSource = flatDataSource
			Catch
				dataAdapter.DataSource = Nothing
			End Try
		End Sub
		Private Sub SetColorizer(ByVal flatDataSource As DashboardFlatDataSource)
			Dim coloringIndices = dashboardItem.Metadata.Arguments.Where(Function(d) d.ColoringMode = ColoringMode.Hue).Select(Function(d) dashboardItem.Metadata.Arguments.IndexOf(d))
			Dim maxcoloringIndex As Integer = 0
			If coloringIndices.Any() Then
				maxcoloringIndex = coloringIndices.Max()
			End If
			sunburst.Colorizer = New SunburstItemColorizer(flatDataSource, maxcoloringIndex)
		End Sub
		Private Sub SetSelectionMode()
			Select Case Interactivity.MasterFilterMode
				Case DashboardItemMasterFilterMode.None
					sunburst.SelectionMode = ElementSelectionMode.None
					Return
				Case DashboardItemMasterFilterMode.Multiple
					sunburst.SelectionMode = ElementSelectionMode.Extended
				Case DashboardItemMasterFilterMode.Single
					sunburst.SelectionMode = ElementSelectionMode.None
			End Select
		End Sub
		Private Sub Sunburst_MouseClick(ByVal sender As Object, ByVal e As MouseEventArgs)
			If Interactivity.MasterFilterMode = DashboardItemMasterFilterMode.Single Then
				Dim hi As SunburstHitInfo = sunburst.CalcHitInfo(e.Location)
				If hi.InSunburstItem AndAlso Not hi.SunburstItem.IsGroup Then
					If Interactivity.CanSetMasterFilter Then
						Interactivity.SetMasterFilter(TryCast(hi.SunburstItem.Tag, DashboardFlatDataSourceRow))
					End If
				End If
			End If
		End Sub
		Private Sub Sunburst_SelectionChanged(ByVal sender As Object, ByVal e As SelectionChangedEventArgs)
			If skipSelectionEvent Then
				Return
			End If
			Dim selectedRows As IEnumerable(Of DashboardFlatDataSourceRow) = e.SelectedItems.OfType(Of DashboardFlatDataSourceRow)()
			If selectedRows.Count() = 0 AndAlso Interactivity.CanClearMasterFilter Then
				Interactivity.ClearMasterFilter()
			ElseIf Interactivity.CanSetMasterFilter Then
				Interactivity.SetMasterFilter(selectedRows)
			End If
		End Sub
		Private Sub ToolTipController_BeforeShow(ByVal sender As Object, ByVal e As ToolTipControllerShowEventArgs)
			Dim item As IHierarchicalItem = TryCast(e.SelectedObject, IHierarchicalItem)
			If item IsNot Nothing Then
				e.ToolTip = String.Format("{0}: {1}", item.Label, multiDimensionalData.GetMeasures()(0).Format(item.Value))
			End If
		End Sub
		Private Sub SunburstLookAndFeelStyleChanged(ByVal sender As Object, ByVal e As EventArgs)
			UpdateSunburstItemStyle()
		End Sub
		Private Sub UpdateSunburstItemStyle()
			Dim currentSkin As Skin = ChartSkins.GetSkin(sunburst.LookAndFeel)
			Dim element As SkinElement = currentSkin(ChartSkins.SkinBackground)
			sunburst.ItemStyle.BorderColor = currentSkin.TranslateColor(element.Color.BackColor)
		End Sub
	End Class
	Friend Class SunburstItemColorizer
		Implements ISunburstColorizer

		Public Event ColorizerChanged As ColorizerChangedEventHandler Implements ISunburstColorizer.ColorizerChanged
		Private defaultColor As Color = Color.Gray
		Private ReadOnly flatData As DashboardFlatDataSource
		Private maxcoloringIndex As Integer
		Public Sub New(ByVal flatData As DashboardFlatDataSource, ByVal maxcoloringIndex As Integer)
			Me.flatData = flatData
			Me.maxcoloringIndex = maxcoloringIndex
		End Sub
		Public Function GetItemColor(ByVal item As ISunburstItem, ByVal group As SunburstItemGroupInfo) As Color Implements ISunburstColorizer.GetItemColor
			If group.GroupLevel < maxcoloringIndex Then
				Dim alpha As Integer = 255 * (group.MaxGroupLevel - group.GroupLevel + 1) / (group.MaxGroupLevel + 1)
				Return Color.FromArgb(alpha, defaultColor)
			End If
			If TypeOf item.Tag Is DashboardFlatDataSourceRow Then
				Dim row As DashboardFlatDataSourceRow = TryCast(item.Tag, DashboardFlatDataSourceRow)
				Dim colorData As Integer = DirectCast(flatData.GetValue(flatData.GetColoringColumn().Name, row), Integer)
				Return Color.FromArgb(colorData)
			End If
			If TypeOf item.Tag Is List(Of Object) Then
				Dim colors As IEnumerable(Of Integer) = (TryCast(item.Tag, List(Of Object))).OfType(Of DashboardFlatDataSourceRow)().Select(Function(row) flatData.GetValue(flatData.GetColoringColumn().Name, row)).Cast(Of Integer)().Distinct()
				If colors.Count() = 1 Then
					Return Color.FromArgb(colors.First())
				End If
			End If
			Return defaultColor
		End Function
	End Class
	Public Class DashboardSunburstControl
		Inherits SunburstControl

		Protected Overrides Sub OnMouseClick(ByVal e As MouseEventArgs)
			If e.Button <> MouseButtons.Right Then
				MyBase.OnMouseClick(e)
			End If
		End Sub
	End Class
End Namespace

File: CS/CustomItemExtension/CustomItems/SunburstChart/SunburstItemControlProvider.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Windows.Forms;
using DevExpress.DashboardCommon;
using DevExpress.DashboardCommon.ViewerData;
using DevExpress.DashboardWin;
using DevExpress.Skins;
using DevExpress.TreeMap;
using DevExpress.Utils;
using DevExpress.Utils.Extensions;
using DevExpress.XtraReports.UI;
using DevExpress.XtraTreeMap;

namespace DevExpress.DashboardWin.CustomItemExtension {
    public class SunburstItemControlProvider : CustomControlProviderBase {
        bool skipSelectionEvent = false;
        SunburstControl sunburst;
        SunburstFlatDataAdapter dataAdapter;
        CustomDashboardItem<SunburstItemMetadata> dashboardItem;
        ToolTipController toolTipController;
        MultiDimensionalData multiDimensionalData;
        DashboardFlatDataSource flatData;
        Title emptyTitle;

        protected override Control Control { get { return sunburst; } }
        public SunburstItemControlProvider(CustomDashboardItem<SunburstItemMetadata> dashboardItem) {
            this.dashboardItem = dashboardItem;
            sunburst = new DashboardSunburstControl();
            sunburst.BorderOptions.Visible = false;
            this.dataAdapter = new SunburstFlatDataAdapter();
            sunburst.DataAdapter = dataAdapter;
            toolTipController = new ToolTipController();
            toolTipController.BeforeShow += ToolTipController_BeforeShow;
            sunburst.ToolTipController = toolTipController;
            sunburst.SelectionChanged += Sunburst_SelectionChanged;
            emptyTitle = new Title() { Visible = false };
            sunburst.Titles.Add(emptyTitle);
            sunburst.MouseClick += Sunburst_MouseClick;
            sunburst.LookAndFeel.StyleChanged += SunburstLookAndFeelStyleChanged;
            UpdateSunburstItemStyle();
        }
        protected override void UpdateControl(CustomItemData customItemData) {
            ClearDataBindings();
            if(ValidateBindings()) {
                flatData = customItemData.GetFlatData(new DashboardFlatDataSourceOptions() { AddColoringColumns = true, AddDisplayTextColumns = true });
                multiDimensionalData = customItemData.GetMultiDimensionalData();
                SetDataBindings(flatData);
                SetColorizer(flatData);
                SetSelectionMode();
            }
        }
        protected override void SetSelection(CustomItemSelection selection) {
            skipSelectionEvent = true;
            IList<DashboardFlatDataSourceRow> selectedRows = selection.GetDashboardFlatDataSourceRows(flatData);
            sunburst.SelectedItems.Clear();
            selectedRows.ForEach(r => sunburst.SelectedItems.Add(r));
            skipSelectionEvent = false;
        }
        protected override XRControl GetPrintableControl(CustomItemData customItemData, CustomItemExportInfo exportInfo) {
            PrintableComponentContainer container = new PrintableComponentContainer();
            container.PrintableComponent = sunburst;
            return container;
        }
        void ClearDataBindings() {
            sunburst.Colorizer = null;
            dataAdapter.DataSource = dataAdapter.ValueDataMember = dataAdapter.LabelDataMember = null;
            dataAdapter.GroupDataMembers.Clear();
        }
        bool ValidateBindings() {
            if(Interactivity.IsDrillDownEnabled) {
                emptyTitle.Text = "The Sunburst Item does not support Drill-Down";
                emptyTitle.Visible = true;
                return false;
            } else
                emptyTitle.Visible = false;
            return dashboardItem.Metadata.Value != null && dashboardItem.Metadata.Arguments.Count() > 0;
        }
        void SetDataBindings(DashboardFlatDataSource flatDataSource) {
            dataAdapter.ValueDataMember = dashboardItem.Metadata.Value.UniqueId;
            dataAdapter.LabelDataMember = flatDataSource.GetDisplayTextColumn(dashboardItem.Metadata.Arguments.Last().UniqueId).Name;
            dataAdapter.GroupDataMembers.AddRange(
                dashboardItem.Metadata.Arguments.Where(d => d != dashboardItem.Metadata.Arguments.Last())
                .Select(d => flatDataSource.GetDisplayTextColumn(d.UniqueId).Name).ToList());
            try {
                dataAdapter.DataSource = flatDataSource;
            } catch {
                dataAdapter.DataSource = null;
            }
        }
        void SetColorizer(DashboardFlatDataSource flatDataSource) {
            var coloringIndices = dashboardItem.Metadata.Arguments.Where(d => d.ColoringMode == ColoringMode.Hue)
                .Select(d => dashboardItem.Metadata.Arguments.IndexOf(d));
            int maxcoloringIndex = 0;
            if(coloringIndices.Any()) maxcoloringIndex = coloringIndices.Max();
            sunburst.Colorizer = new SunburstItemColorizer(flatDataSource, maxcoloringIndex);
        }
        void SetSelectionMode() {
            switch(Interactivity.MasterFilterMode) {
                case DashboardItemMasterFilterMode.None:
                    sunburst.SelectionMode = ElementSelectionMode.None;
                    return;
                case DashboardItemMasterFilterMode.Multiple:
                    sunburst.SelectionMode = ElementSelectionMode.Extended;
                    break;
                case DashboardItemMasterFilterMode.Single:
                    sunburst.SelectionMode = ElementSelectionMode.None;
                    break;
            }
        }
        void Sunburst_MouseClick(object sender, MouseEventArgs e) {
            if(Interactivity.MasterFilterMode == DashboardItemMasterFilterMode.Single) {
                SunburstHitInfo hi = sunburst.CalcHitInfo(e.Location);
                if(hi.InSunburstItem && !hi.SunburstItem.IsGroup) {
                    if(Interactivity.CanSetMasterFilter)
                        Interactivity.SetMasterFilter(hi.SunburstItem.Tag as DashboardFlatDataSourceRow);
                }
            }
        }
        void Sunburst_SelectionChanged(object sender, SelectionChangedEventArgs e) {
            if(skipSelectionEvent) return;
            IEnumerable<DashboardFlatDataSourceRow> selectedRows = e.SelectedItems.OfType<DashboardFlatDataSourceRow>();
            if(selectedRows.Count() == 0 && Interactivity.CanClearMasterFilter)
                Interactivity.ClearMasterFilter();
            else if(Interactivity.CanSetMasterFilter)
                Interactivity.SetMasterFilter(selectedRows);
        }
        void ToolTipController_BeforeShow(object sender, ToolTipControllerShowEventArgs e) {
            IHierarchicalItem item = e.SelectedObject as IHierarchicalItem;
            if(item != null)
                e.ToolTip = string.Format("{0}: {1}", item.Label, multiDimensionalData.GetMeasures()[0].Format(item.Value));
        }
        void SunburstLookAndFeelStyleChanged(object sender, EventArgs e) {
            UpdateSunburstItemStyle();
        }
        void UpdateSunburstItemStyle() {
            Skin currentSkin = ChartSkins.GetSkin(sunburst.LookAndFeel);
            SkinElement element = currentSkin[ChartSkins.SkinBackground];
            sunburst.ItemStyle.BorderColor = currentSkin.TranslateColor(element.Color.BackColor);
        }
    }
    class SunburstItemColorizer : ISunburstColorizer {
        public event ColorizerChangedEventHandler ColorizerChanged;
        Color defaultColor = Color.Gray;
        readonly DashboardFlatDataSource flatData;
        int maxcoloringIndex;
        public SunburstItemColorizer(DashboardFlatDataSource flatData, int maxcoloringIndex) {
            this.flatData = flatData;
            this.maxcoloringIndex = maxcoloringIndex;
        }
        public Color GetItemColor(ISunburstItem item, SunburstItemGroupInfo group) {
            if(group.GroupLevel < maxcoloringIndex) {
                int alpha = 255 * (group.MaxGroupLevel - group.GroupLevel + 1) / (group.MaxGroupLevel + 1);
                return Color.FromArgb(alpha, defaultColor);
            }
            if(item.Tag is DashboardFlatDataSourceRow) {
                DashboardFlatDataSourceRow row = item.Tag as DashboardFlatDataSourceRow;
                int colorData = (int)flatData.GetValue(flatData.GetColoringColumn().Name, row);
                return Color.FromArgb(colorData);
            }
            if(item.Tag is List<object>) {
                IEnumerable<int> colors = (item.Tag as List<object>).OfType<DashboardFlatDataSourceRow>()
                    .Select(row => flatData.GetValue(flatData.GetColoringColumn().Name, row)).Cast<int>().Distinct();
                if(colors.Count() == 1)
                    return Color.FromArgb(colors.First());
            }
            return defaultColor;
        }
    }
    public class DashboardSunburstControl : SunburstControl {
        protected override void OnMouseClick(MouseEventArgs e) {
            if(e.Button != MouseButtons.Right)
                base.OnMouseClick(e);
        }
    }
}
